File: agent/foundry_agent.py
"""
Azure AI Foundry Agent with MCP Tools + Bing Grounding

Connects to our Philly poverty profiteering MCP server (deployed on Container Apps)
and optionally adds Bing web search grounding for real-time data.

Prerequisites:
  1. Azure AI Foundry project (https://ai.azure.com)
  2. Container App running the MCP server (deploy with infra/deploy-agent.sh)
  3. (Optional) Grounding with Bing Search resource connected to the project

Environment variables:
  PROJECT_ENDPOINT           - Azure AI Foundry project endpoint
  MODEL_DEPLOYMENT_NAME      - Model deployment (default: gpt-4o)
  MCP_SERVER_URL             - Container App MCP endpoint (e.g. https://philly-mcp-server.*.azurecontainerapps.io/mcp)
  BING_CONNECTION_NAME       - (Optional) Bing grounding connection ID for web search

Usage:
  pip install -r requirements.txt
  az login
  python foundry_agent.py
  python foundry_agent.py --query "Who are the top violators in 19134?"
"""

import argparse
import os
import sys
import time

from azure.ai.projects import AIProjectClient
from azure.identity import DefaultAzureCredential
from azure.ai.agents.models import (
    BingGroundingTool,
    ListSortOrder,
    McpTool,
    MessageTextContent,
    RequiredMcpToolCall,
    SubmitToolApprovalAction,
    ToolApproval,
)


def create_agent(project_client: AIProjectClient):
    """Create a Foundry agent with MCP + optional Bing tools."""
    mcp_url = os.environ.get("MCP_SERVER_URL")
    if not mcp_url:
        print("ERROR: MCP_SERVER_URL environment variable is required.")
        print("  Set it to your Container App MCP endpoint, e.g.:")
        print("  export MCP_SERVER_URL=https://philly-mcp-server.<region>.azurecontainerapps.io/mcp")
        sys.exit(1)

    model = os.environ.get("MODEL_DEPLOYMENT_NAME", "gpt-4o")

    # MCP tool — connects to our Philly property data server
    mcp_tool = McpTool(
        server_label="philly-stats",
        server_url=mcp_url,
    )

    # Combine tool definitions
    all_tools = mcp_tool.definitions

    # Optional: Bing grounding for web search
    bing_conn = os.environ.get("BING_CONNECTION_NAME")
    bing_tool = None
    if bing_conn:
        bing_tool = BingGroundingTool(connection_id=bing_conn)
        all_tools = all_tools + bing_tool.definitions
        print(f"  Bing grounding enabled (connection: ...{bing_conn[-20:]})")
    else:
        print("  Bing grounding disabled (set BING_CONNECTION_NAME to enable)")

    agent = project_client.agents.create_agent(
        model=model,
        name="philly-investigator",
        instructions=(
            "You are an investigative analyst specializing in Philadelphia property data. "
            "You have access to tools that query a database of ~29 million rows covering "
            "property ownership networks, code violations, demolitions, business licenses, "
            "and tax assessments. Use these tools to identify patterns of neglect, "
            "exploitative landlords, and poverty profiteering. "
            "When answering, cite specific data (parcel numbers, violation counts, addresses). "
            "If Bing web search is available, use it to find recent news about properties or owners."
        ),
        tools=all_tools,
    )
    print(f"  Created agent: {agent.id}")
    return agent, mcp_tool, bing_tool


def run_query(project_client: AIProjectClient, agent, mcp_tool, bing_tool, query: str):
    """Run a single query against the agent and print the response."""
    agents = project_client.agents

    # Create thread and message
    thread = agents.threads.create()
    agents.messages.create(thread_id=thread.id, role="user", content=query)

    # Build tool resources — auto-approve MCP calls (we trust our own server)
    mcp_tool.update_headers("X-Source", "foundry-agent")
    tool_resources = mcp_tool.resources

    run = agents.runs.create(
        thread_id=thread.id,
        agent_id=agent.id,
        tool_resources=tool_resources,
    )

    # Poll until complete
    while run.status in ("queued", "in_progress", "requires_action"):
        time.sleep(1)
        run = agents.runs.get(thread_id=thread.id, run_id=run.id)

        # Auto-approve MCP tool calls
        if run.status == "requires_action" and isinstance(
            run.required_action, SubmitToolApprovalAction
        ):
            tool_calls = run.required_action.submit_tool_approval.tool_calls
            approvals = []
            for tc in tool_calls:
                if isinstance(tc, RequiredMcpToolCall):
                    print(f"  [MCP] {tc.name}({tc.arguments[:80]}...)")
                    approvals.append(
                        ToolApproval(
                            tool_call_id=tc.id,
                            approve=True,
                            headers=mcp_tool.headers,
                        )
                    )
            if approvals:
                agents.runs.submit_tool_outputs(
                    thread_id=thread.id,
                    run_id=run.id,
                    tool_approvals=approvals,
                )

    if run.status == "failed":
        print(f"\nRun failed: {run.last_error}")
        return

    # Print assistant response
    messages = agents.messages.list(thread_id=thread.id, order=ListSortOrder.ASCENDING)
    for msg in messages:
        if msg.role == "assistant" and msg.text_messages:
            text = msg.text_messages[-1].text.value
            print(f"\n{text}")

    # Cleanup thread
    agents.threads.delete(thread_id=thread.id)


def interactive_loop(project_client: AIProjectClient, agent, mcp_tool, bing_tool):
    """Run an interactive chat loop."""
    print("\nPhilly Property Investigator (type 'quit' to exit)")
    print("-" * 50)

    while True:
        try:
            query = input("\nYou: ").strip()
        except (EOFError, KeyboardInterrupt):
            break

        if not query or query.lower() in ("quit", "exit", "q"):
            break

        run_query(project_client, agent, mcp_tool, bing_tool, query)

    print("\nGoodbye.")


def main():
    parser = argparse.ArgumentParser(description="Philly Property Investigator Agent")
    parser.add_argument("--query", "-q", help="Single query (skip interactive mode)")
    args = parser.parse_args()

    project_endpoint = os.environ.get("PROJECT_ENDPOINT")
    if not project_endpoint:
        print("ERROR: PROJECT_ENDPOINT environment variable is required.")
        print("  Set it to your Azure AI Foundry project endpoint.")
        sys.exit(1)

    print("Initializing Foundry agent...")
    project_client = AIProjectClient(
        endpoint=project_endpoint,
        credential=DefaultAzureCredential(),
    )

    with project_client:
        agent, mcp_tool, bing_tool = create_agent(project_client)

        try:
            if args.query:
                run_query(project_client, agent, mcp_tool, bing_tool, args.query)
            else:
                interactive_loop(project_client, agent, mcp_tool, bing_tool)
        finally:
            project_client.agents.delete_agent(agent.id)
            print("Agent cleaned up.")


if __name__ == "__main__":
    main()

File: mcp-server/src/foundry-agent.ts
/**
 * Foundry Agent — Assistants API lifecycle.
 * Azure manages the tool-calling loop and threads persist server-side.
 * We just execute tools when the run enters "requires_action" state.
 */

import { AzureOpenAI } from "openai";
import { DefaultAzureCredential, getBearerTokenProvider } from "@azure/identity";
import { SYSTEM_PROMPT, TOOLS, executeTool } from "./tool-executor.js";

const AZURE_OPENAI_ENDPOINT = process.env.AZURE_OPENAI_ENDPOINT ?? "https://foundry-og-agents.cognitiveservices.azure.com/";
const AZURE_OPENAI_API_VERSION = "2025-01-01-preview";
const AGENT_MODEL = "gpt-5";
const AGENT_NAME = "philly-investigator";

let clientInstance: AzureOpenAI | null = null;
let agentId: string | null = null;

function getClient(): AzureOpenAI {
  if (!clientInstance) {
    const credential = new DefaultAzureCredential();
    const tokenProvider = getBearerTokenProvider(credential, "https://cognitiveservices.azure.com/.default");
    clientInstance = new AzureOpenAI({
      azureADTokenProvider: tokenProvider,
      endpoint: AZURE_OPENAI_ENDPOINT,
      apiVersion: AZURE_OPENAI_API_VERSION,
    });
  }
  return clientInstance;
}

/**
 * Ensure the assistant exists. Searches for an existing one by name,
 * creates it if not found. Caches the ID in memory.
 */
export async function ensureAgent(): Promise<string> {
  if (agentId) return agentId;

  const client = getClient();

  // Search for existing assistant by name
  const list = await client.beta.assistants.list({ limit: 100 });
  for (const assistant of list.data) {
    if (assistant.name === AGENT_NAME) {
      agentId = assistant.id;
      console.log(`[agent] Found existing assistant: ${agentId}`);
      return agentId;
    }
  }

  // Create new assistant with our tools
  const assistant = await client.beta.assistants.create({
    name: AGENT_NAME,
    description: "Investigates poverty profiteering patterns in Philadelphia using 29M rows of public data.",
    model: AGENT_MODEL,
    instructions: SYSTEM_PROMPT,
    tools: TOOLS.filter((t): t is typeof t & { type: "function"; function: object } => t.type === "function")
      .map(t => ({
        type: "function" as const,
        function: t.function,
      })),
    temperature: 0.7,
  });

  agentId = assistant.id;
  console.log(`[agent] Created new assistant: ${agentId}`);
  return agentId;
}

/** Create a new thread. Returns the thread ID. */
export async function createThread(): Promise<string> {
  const client = getClient();
  const thread = await client.beta.threads.create();
  console.log(`[agent] Created thread: ${thread.id}`);
  return thread.id;
}

export interface AgentResponse {
  reply: string;
  toolCalls: Array<{ name: string; args: Record<string, unknown> }>;
}

/** Send a message to a thread and run the assistant. Polls until complete. */
export async function sendMessage(threadId: string, message: string): Promise<AgentResponse> {
  const client = getClient();
  const assistantId = await ensureAgent();
  const toolCalls: Array<{ name: string; args: Record<string, unknown> }> = [];

  // Add the user message to the thread
  await client.beta.threads.messages.create(threadId, {
    role: "user",
    content: message,
  });

  // Create a run
  let run = await client.beta.threads.runs.create(threadId, {
    assistant_id: assistantId,
  });

  console.log(`[agent] Run ${run.id} started (status: ${run.status})`);

  // Poll until terminal state
  while (true) {
    if (run.status === "completed") {
      break;
    }

    if (run.status === "failed") {
      const errorMsg = run.last_error?.message ?? "Unknown error";
      throw new Error(`Agent run failed: ${errorMsg}`);
    }

    if (run.status === "cancelled" || run.status === "expired") {
      throw new Error(`Agent run ${run.status}`);
    }

    if (run.status === "requires_action") {
      const requiredAction = run.required_action;
      if (requiredAction?.type === "submit_tool_outputs") {
        const toolOutputs = [];
        for (const tc of requiredAction.submit_tool_outputs.tool_calls) {
          const args = JSON.parse(tc.function.arguments);
          toolCalls.push({ name: tc.function.name, args });

          console.log(`  [agent-tool] ${tc.function.name}(${JSON.stringify(args).slice(0, 100)})`);
          const result = await executeTool(tc.function.name, args);

          toolOutputs.push({
            tool_call_id: tc.id,
            output: result,
          });
        }

        // Submit tool outputs and get updated run
        run = await client.beta.threads.runs.submitToolOutputs(run.id, {
          thread_id: threadId,
          tool_outputs: toolOutputs,
        });
        console.log(`[agent] Submitted tool outputs (status: ${run.status})`);
        continue;
      }
    }

    // Wait before polling again
    await new Promise(resolve => setTimeout(resolve, 1000));
    run = await client.beta.threads.runs.retrieve(run.id, { thread_id: threadId });
  }

  // Extract the assistant's response from thread messages
  const messages = await client.beta.threads.messages.list(threadId, {
    order: "desc",
    limit: 1,
  });

  let reply = "(no response)";
  const lastMessage = messages.data[0];
  if (lastMessage?.role === "assistant" && lastMessage.content.length > 0) {
    const textBlock = lastMessage.content.find(b => b.type === "text");
    if (textBlock?.type === "text") {
      reply = textBlock.text.value;
    }
  }

  return { reply, toolCalls };
}
